Padding: support complex dtypes

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50594 Padding: support complex dtypes**
* #50593 Improve complex support in common_nn test machinery

Fixes #50234

Differential Revision: [D25987316](https://our.internmc.facebook.com/intern/diff/D25987316)

peterbell10 committed:
Padding: support complex dtypes

[ghstack-poisoned]

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -798,14 +797,13 @@ def test_stft_roundtrip_complex_window(self, device, dtype):
              torch.randn(807, device=device, dtype=dtype),
-             torch.randn(12, 14, device=device, dtype=dtype),
-             torch.randn(9, 6, device=device, dtype=dtype)),
+             torch.randn(12, 60, device=device, dtype=dtype)),
```
Why change these inputs?

peterbell10 replied:
reflect mode needs to have enough data to reflect, otherwise you get an error:
```
RuntimeError: Argument #4: Padding size should be less than the corresponding input dimension, but got: [...]
```

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -798,14 +797,13 @@ def test_stft_roundtrip_complex_window(self, device, dtype):
             (None, 10),
             # pad_mode
-            ("constant",),
+            ("constant", "reflect", "circular"),
```
What about replicate? It's OK if we don't support replicate in this PR, but then let's update the documentation to clarify which modes stft supports when given complex inputs and file a follow-up task for replicate padding.

mruberry reviewed (commented):
Hey @peterbell10! Overall this looks great and very thorough. I've asked @anjali411 to take a look, too. I have two questions inline that I'm curious to hear your thinking on.

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234


[ghstack-poisoned]

peterbell10 commented:
@mrubery PTAL, added the padding mode missing from the tests. Also fixed merge conflicts with #50632.

As an aside, is there anything we can do to reducing merge conflicts in the `GRADIENT_IMPLEMENTED_FOR_COMPLEX` list? I've had multiple conflicts for every PR that adds complex gradient support. Perhaps one entry per line and alpha-sorting would help.

mruberry commented:
I'll take a look ASAP.

> As an aside, is there anything we can do to reducing merge conflicts in the `GRADIENT_IMPLEMENTED_FOR_COMPLEX` list? I've had multiple conflicts for every PR that adds complex gradient support. Perhaps one entry per line and alpha-sorting would help.

Good point. What do you think, @anjali411?

anjali411 reviewed (approved):
LGTM! 

> As an aside, is there anything we can do to reducing merge conflicts in the GRADIENT_IMPLEMENTED_FOR_COMPLEX list? I've had multiple conflicts for every PR that adds complex gradient support. Perhaps one entry per line and alpha-sorting would help.

@peterbell10 yeah I agree that merge conflicts are pretty annoying however I think there'd still be merge conflicts with those solutions no?

peterbell10 commented:
@anjali411 two PRs appending to the same list are guaranteed to create merge conflicts. Inserting into the middle of an alpha-sorted list only creates conflicts if the operators are next to each other in the alphabet. So yes there can still be merge conflicts, but at least some of the time it can be merged automatically.

anjali411 commented:
> @anjali411 two PRs appending to the same list are guaranteed to create merge conflicts. Inserting into the middle of an alpha-sorted list only creates conflicts if the operators are next to each other in the alphabet. So yes there can still be merge conflicts, but at least some of the time it can be merged automatically.

makes sense okay!

anjali411 committed:
Update on "Padding: support complex dtypes"


Fixes #50234

Differential Revision: [D25987316](https://our.internmc.facebook.com/intern/diff/D25987316)

[ghstack-poisoned]

anjali411 commented:
One of the caffe2 tests error out:
```
    ✗ Fail: caffe2/test:jit - test_nn_module_tests (jit.test_complexity.TestComplexity) (6.023)
Test output:
> RuntimeError: ('User provided tensor is real for a test that runs with complex dtype, ', 'which is not supported for now')
  File "/usr/local/fbcode/platform007/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/fbcode/platform007/lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/data/users/chourdiaanjali/fbsource/fbcode/buck-out/opt/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/common_utils.py", line 595, in wrapper
    fn(*args, **kwargs)
  File "/data/users/chourdiaanjali/fbsource/fbcode/buck-out/opt/gen/caffe2/test/jit#binary,link-tree/jit/test_complexity.py", line 73, in test_nn_module_tests
    out = try_get_nn_module_compiled_mod_and_inputs(**test)
  File "/data/users/chourdiaanjali/fbsource/fbcode/buck-out/opt/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/jit_metaprogramming_utils.py", line 546, in try_get_nn_module_compiled_mod_and_inputs
    args_variable, kwargs_variable = create_input(input)
  File "/data/users/chourdiaanjali/fbsource/fbcode/buck-out/opt/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/common_methods_invocations.py", line 2692, in create_input
    args_out = tuple(map_arg(arg) for arg in call_args)
  File "/data/users/chourdiaanjali/fbsource/fbcode/buck-out/opt/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/common_methods_invocations.py", line 2692, in <genexpr>
    args_out = tuple(map_arg(arg) for arg in call_args)
  File "/data/users/chourdiaanjali/fbsource/fbcode/buck-out/opt/gen/caffe2/test/jit#binary,link-tree/torch/testing/_internal/common_methods_invocations.py", line 2683, in map_arg
    "which is not supported for now")
stdout:
Couldn't download test skip set, leaving all tests enabled...

stderr:


Summary
  Fail: 1
    ✗ caffe2/test:jit - test_nn_module_tests (jit.test_complexity.TestComplexity)
```

peterbell10 committed:
Update on "Padding: support complex dtypes"


Fixes #50234

Differential Revision: [D25987316](https://our.internmc.facebook.com/intern/diff/D25987316)

[ghstack-poisoned]
